src/nodes.coffee

```diff
@@ -1814,6 +1847,80 @@ exports.Assign = class Assign extends Base
     else
       answer
 
+  # Check object destructuring variable for rest elements;
+  # can be removed once ES proposal hits Stage 4.
+  compileObjectDestruct: (o) ->
+    # Per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assignment_without_declaration,
+    # if we’re destructuring without declaring, the destructuring assignment must be wrapped in parentheses.
+    # ({a, b} = obj)
+    # Helper function setScopeVar() declares vars 'a' and 'b' at the top of the current scope.
+    setScopeVar = (prop) ->
+      newVar = false
+      return if prop instanceof Assign and prop.value.base instanceof Obj
+      if prop instanceof Assign
+        if prop.value.base instanceof IdentifierLiteral
+          newVar = prop.value.base.compile o
+        else 
+          newVar = prop.variable.base.compile o
+      else
+        newVar = prop.compile o
+      o.scope.add(newVar, 'var', true) if newVar
+    # Helper function getPropVaues() returns compiled object property value.
+    # These values are then passed as an argument to helper function objectWithoutKeys 
+    # which is used to assign object value to the destructuring rest variable.
+    getPropValue = (prop) ->
+      wrapInQutes = (prop) -> (new Literal "'#{prop.compile o}'").compile o
+      setScopeVar prop # Declare a variable in the scope.
+      if prop instanceof Assign
+        return prop.variable.compile o if prop.variable.base instanceof StringWithInterpolations or prop.variable.base instanceof StringLiteral
+        return wrapInQutes prop.variable
+      else
+        return wrapInQutes prop
+    # Recursive function for searching and storing rest elements in objects.
+    # Parameter props[] is used to store nested object properties,
+    # e.g. `{a: {b, c: {d, r1...}, r2...}, r3...} = obj`.
+    traverseRest = (properties, path = []) ->
+      results = []
+      restElement = no
+      for prop, key in properties
+        if prop instanceof Assign and prop.value.base instanceof Obj          
+          results = traverseRest prop.value.base.objects, [path..., getPropValue prop]
+        if prop instanceof Splat
+          prop.error "multiple rest elements are disallowed in object destructuring" if restElement
+          restElement = {
+            key,
+            name: prop.unwrap(),
+            props: ((new Literal "[#{p}]").compile(o) for p in path).join ""            
+          }  
+      if restElement
+        # Remove rest element from the object.
+        properties.splice restElement.key, 1
+        # Prepare array of compiled property keys to be excluded from the object.
+        restElement["excludeProps"] = new Literal "[#{(getPropValue(prop) for prop in properties)}]"
```

Comment:
Because currently, we're not forcing splat to be the last element in the object destructuring, the ["key"] is only used for storing the position of splat property in the object's properties list.
(I think it had also some other use in the previous iterations.)
Anyway, your right, it would be cleaner if a simple variable is used instead.